python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+        self._log("Waiting for %d background threads to stop..." % len(self._all_threads))
+        for thread in self._all_threads:
+            thread.shut_down()
+            #thread.deleteLater()
+            #thread = None
+            #thread.stop(wait_for_completion = True)
+        self._log("Shut down successfully!")
+        self._all_threads = []
+
+    def add_task(self, cbl, priority=None, group=None, upstream_task_ids=None, args=[], **kwargs):
+        """
+        Add a new task to the queue.  A task is a callable method/class together with any arguments that
+        should be passed to the callable when it is called.
+
+        :param cbl:                 The callable function/class to call when executing the task
+        :param priority:            The priority this task should be run with.  Tasks with higher priority
+                                    are run first.
+        :param group:               The group this task belongs to.  Task groups can be used to simplify task
+                                    management (e.g. stop a whole group, be notified when a group is complete)
+        :param upstream_task_ids:   A list of any upstream tasks that should be completed before this task
+                                    is run.  The results from any upstream tasks are appended to the kwargs
+                                    for this task.
+        :param args:                A list of unnamed parameters to be passed to the callable when running the 
+                                    task
+        :param **kwargs:            A dictionary of named parameters to be passed to the callable when running 
+                                    the task
+        :returns:                   A unique id representing the task.
+        """
+        if not callable(cbl):
+            raise TankError("The task function, method or object '%s' must be callable!" % cbl)
+
+        upstream_task_ids = set(upstream_task_ids or [])
+
+        # create a new task instance:
+        task_id = self._next_task_id
+        self._next_task_id += 1
+        new_task = _BackgroundTask(task_id, cbl, group, priority, args, **kwargs)
+
+        # add the task to the pending queue:
+        self._pending_tasks_by_priority.setdefault(priority, []).append(new_task)
+
+        # add tasks to various look-ups:
+        self._tasks_by_id[new_task.uid] = new_task
+        self._group_task_map.setdefault(group, set()).add(new_task.uid)
+
+        # keep track of the task dependencies:
+        self._upstream_task_map[new_task.uid] = upstream_task_ids
+        for us_task_id in upstream_task_ids:
+            self._downstream_task_map.setdefault(us_task_id, set()).add(new_task.uid)
+
+        self._log("Added Task %s to the queue" % new_task)
+
+        # and start the next task:
+        self._start_tasks()
+
+        return new_task.uid
+
+    def add_pass_through_task(self, priority=None, group=None, upstream_task_ids=None, **kwargs):
```

Comment:
If this kind of task is meant for synchronization, maybe add_synchronization_task would be a better name. Are there any other uses for this task type?